File: solutions/986. Interval List Intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        #soln 0 - first attempt, Time O(min(N, M)), Space O(min(N,M))
        i, j = 0, 0
        ans = []
        while i < len(firstList) and j < len(secondList):
            A, B = firstList[i], secondList[j]
            if not (A[1] < B[0] or A[0] > B[1]):
                #there is an intersection
                ans.append([max(A[0], B[0]), min(A[1], B[1])])
            if A[1] <= B[1]:
                i += 1
            else:
                j += 1
        return ans
        

File: solutions/986. Interval List Intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        // time O(N+M), space O(1)
        int i=0, j=0;
        vector<vector<int>> ans;
        
        while (i<firstList.size() && j<secondList.size()) {
            int s1=firstList[i][0], e1=firstList[i][1], s2=secondList[j][0], e2=secondList[j][1];
            if (s2 <= e1 && s1 <= e2) {
                ans.push_back({max(s1, s2), min(e1, e2)});
            }
            
            if (e1 <= e2) i++;
            else j++;
        }
        return ans;
    }
};
